Azure Blob Storage Disks

jkuklis opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Adding support for disks backed by Azure Blob Storage, in a similar way it has been done for disks backed by AWS S3. Current implementation allows for all the basic disk operations.


# Detailed description

## Context

At Contentsquare, we use S3 Disks in our AWS servers for example for storing raw data, or certain metadata used for data loss monitoring - for these use cases regular disks would be too expensive. The `DiskS3` approach (and not e.g. `S3Engine` one) is the one which fits us best, as it can be used with `MergeTree`s.

Soon we will need a similar solution in Azure servers. We decided internally that the best strategy for us is to develop the same kind of logic for Blob Storage Disks. We raised an issue describing our approach in September: https://github.com/ClickHouse/ClickHouse/issues/29430.

## Implementation

The code is not fully polished yet, but we would like to start discussing the project.

We have done the following parts:
- Azure SDK dependency
- Blob Storage Disk implementation
- Disk registration
- Read and Write Buffers
- Deployment and tests with production traffic (still experiencing some complex errors that we would like to discuss)

### Azure SDK dependency

We added `azure-sdk-for-cpp` directory and `azure-cmake` directory with custom `CMakeLists.txt` to `contrib`, updated necessary `.cmake` and `CMakeLists.txt` files. We also had to add a (deprecated) source file to `borignssl-cmake` `CMakeLists.txt`, as one of its functions is used in the Azure SDK (at least it was at the time of development).

### Blob Storage Disk implementation

Implemented main operations: `readFile`, `writeFile`, `checkUniqueId`, `removeFromRemoteFS` and helpers. A lot of operations like `moveFile` or `createHardLink` come for free from `IDiskRemote` (unlike in S3 Disks, no custom metadata is created for Blob Storage, at least yet).

### Disk registration

Following the same patterns as in S3 Disk registration, a local metadata directory is created, simple tests are run, and files are cached if cache is enabled. Only a single authentication method is supported (and expected) at the moment: Managed Identity Credential, which is a form of instance-assigned auth.

### Read and Write Buffers

Tricky part to get right, but both buffers seem to operate as intended and use the Azure API efficiently. This part should be reviewed very carefully, I am the least confident in it.

### Deployment and tests with production traffic

We have tested the server with Yandex.Metrica datasets, we weren't able to spot any remaining errors there. We have also deployed the servers in a cluster receiving our production traffic, these ran there for a couple of days. Here are the findings:

- all the servers have run without a fatal crash,
- there have been quite a lot of `NETWORK_ERROR` and `SOCKET_TIMEOUT` errors spread across the logs, I don't know yet if that is caused by our pipeline,
- in some shards we started receiving Azure `The specified blob does not exist` errors - somehow the `MergeTree` metadata kept trace of files which either never existed in Blob Storage or were deleted earlier,
- in one shard we've got `FILE_DOESNT_EXIST` error a couple of times, again as if there was some mismatch in `MergeTree` metadata, 
- before errors started popping up more frequently, the servers were able to collect almost all of the traffic (in one shard `153539` rows vs expected `153544` for the first day with `FINAL`), however, we see quite a lot of duplicated rows without counterparts with sign -1 (for the same day `205690` rows in total, `203583` with sign +1).

## Further steps

On our side, we will continue to investigate the remaining errors. These, however, are not easy to track, as they appear only once the full system is deployed. If you have any insights about the errors we have encountered, please let us know, we are also happy to share more details and discuss.

We have several questions on our own, including the lines marked with `TODO` and `NOTE` in the code - we left them there on purpose, to be able to discuss them. Other questions include:

- What is the main reason for having a POCO client for S3, considering AWS already provides HTTP clients using e.g. libcurl?
- What is the purpose of the proxy configuration part for S3?
- We only provide a single authentication method for now, as it is not easy for us to test other methods, can we leave like it that for the first iteration of the project?
- How does multithreading impact this code? Do we need to worry about possible race conditions?
- Have we missed any vital components of the Disk implementation?
- How to deal with the remaining copy-paste from S3?
- How to better unit-test this code? Some very basic tests are run during Disk registration, but we do not feel it is enough.

\
\
Thanks for attention, let us know what you think!

Jakub Kuklis
Contentsquare

jkuklis committed:
Initial working example of Azure Blob Storage manipulation from within ClickHouse

jkuklis committed:
Using internal curl library

jkuklis committed:
Cleaning azure-cmake CMakeLists.txt

jkuklis committed:
Removing unused file

jkuklis committed:
Adding a flag to use Azure Blob Storage

jkuklis committed:
Skeleton for DiskBlobStorage

jkuklis committed:
Giving Disk Blob Storage files proper names

jkuklis committed:
Registering the Blob Storage Disk

jkuklis committed:
Forgot about commiting this file

jkuklis committed:
Moving implementation to the .cpp file

jkuklis committed:
The following query passes without erors: CREATE TABLE ... ENGINE = MergeTree ... storage_policy = 'blob_storage', debug constants present in the code

jkuklis committed:
Adding an authenticated client to DiskBlobStorage

jkuklis committed:
Starting insertion pipeline

jkuklis committed:
Progressing with WriteBufferFromBlobStorage

jkuklis committed:
Uploading to Blob Storage on INSERT seems to do something reasonable

jkuklis committed:
Managed to read markers file in SELECT, getting error for CODEC version now

jkuklis committed:
Add azure-sdk-for-cpp as submodule in contrib/azure, linking the main repo here instead of a fork, needs to be updated eventually according to https://clickhouse.com/docs/en/development/contrib/#adding-third-party-libraries

jkuklis committed:
Patch for fixing warnings triggered by Azure SDK, cd to contrib/azure and use 'git apply ../azure-cmake/azure_patch

youennL-cs committed:
Fix compilation issues with Azure in order to follow the main git and not a patched one.

jkuklis committed:
Initial check with write and read works for the Blob Storage Disk

jkuklis committed:
Basic inserts and selects work for a MergeTree table with a Blob Disk, based on a single column table

jkuklis committed:
Two column MergeTree table with Blob Storage disk work! Fixed buffer in ReadBufferFromBlobStorage to be permanent (class attribute)

jkuklis committed:
Code cleanup, leaving some debug information printing under #ifdef VERBOSE_DEBUG_MODE

jkuklis committed:
Increase default thread pool size

jkuklis committed:
Enabling local cache for DiskBlobStorage

jkuklis committed:
Adding settings getter and applyNewSettings

jkuklis committed:
Adding multipart upload to WriteBuffer, retries in ReadBuffer, additional settings for buffers

jkuklis committed:
Blob Storage endpoint url verification using re2

jkuklis committed:
Change false to true in supportZeroCopyReplication

jkuklis committed:
ReadBufferFromBlobStorage operating like a proper stream, rather than a single download-read buffer

jkuklis committed:
checkUniqueId implementation using ListBlobs

jkuklis committed:
File deletion prototype

jkuklis committed:
Further code cleanup and error codes improvement

jkuklis committed:
Checking for success of file removal operation

jkuklis committed:
Shared pointer for Azure client and fixing path reference in ReadBuffer

jkuklis committed:
Adding finalize() to WriteBuffer, always using block blob uploads, code cleanup

jkuklis committed:
Better search for blobs with a given name

jkuklis committed:
Changing offset in ReadBuffer to size_t

jkuklis committed:
Minor improvements

jkuklis committed:
Using unique_ptr for Blob Client in Read Buffer

jkuklis committed:
Bringing back old single part upload limit

jkuklis committed:
Maximum read

jkuklis committed:
Fixing read size error plus minor improvements

jkuklis committed:
Adding a new test for read with offset

jkuklis committed:
Code cleanup

jkuklis committed:
Catch Blob Storage exceptions

Commit:
Small changes following review

jkuklis committed:
Minor refactors before PR

jkuklis committed:
Rebase on master

jkuklis committed:
Refactors after proper rebase on master

jkuklis committed:
Adding max_retries config

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=31505) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=31505) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: jkuklis<br/>:white_check_mark: youennL-cs<br/>:x: r-zenine<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=31505) it.</sub>

kssenii commented:
:+1:
Let'd add integration tests in `tests/integration`.
See `tests/integration/test_merge_tree_s3`, `tests/integration/test_merge_tree_hdfs`

ianton-ru commented on src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:
```diff
@@ -8,6 +8,10 @@
 #include <IO/ReadBufferFromS3.h>
 #endif
 
+#if USE_AWS_S3
```
#if USE_AZURE_BLOB_STORAGE

jkuklis replied:
Fixed

kssenii commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,164 @@
+
+off_t ReadBufferFromBlobStorage::getPosition()
+{
+    // TODO: why is it `return offset - available()` in S3?
```
Because in `nextImpl()` `offset` is incremented with `bytes_read`, but we actually did not yet read it, so need to subtract `available()`.

jkuklis replied:
Hm, considering your explanation, wouldn't it rather be something like `return offset()` or `return impl->offset()` then (underneath `pos - working_buffer.begin()`, like in ReadBufferFromMemory.h) ?
Using `offset - available()`, one gets `offset - (working_buffer.end() - pos) = offset + pos - working_buffer.end()`, which doesn't make sense to me.
The explanation for `getPosition()` is "Offset from the begin of the underlying buffer / file corresponds to the buffer current position", but with the initial `offset = 0` and `pos = working_buffer.begin()` we get `getPosition() = (-1) * working_buffer.size()`.

kssenii replied:
>  but with the initial offset = 0 and pos = working_buffer.begin() we get getPosition() = (-1) * working_buffer.size()

No, this will not happen, because with `offset = 0`, i.e. before we read anything we have `working_buffer.begin() == working_buffer.end()` so `getPosition()` will return `0`.

jkuklis replied:
In this case I should use `offset - available()` here too, right? I also increment `offset` with `bytes_read` after data fetch, but before a read.

kssenii commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,164 @@
+    catch (const Azure::Storage::StorageException & e)
+    {
+        LOG_INFO(log, "Exception caught during Azure Download for file {} at offset {} : {}", path, offset, e.Message);
+        throw e;
```
Do we need retries?
Might be useful in case of setting `remote_filesystem_read_method=threadpool`, because there is some possibility to get something like too many requests error.

jkuklis replied:
Added retries.

ianton-ru commented on src/Disks/BlobStorage/registerDiskBlobStorage.cpp:
```diff
@@ -0,0 +1,173 @@
+
+void validate_endpoint_url(const String & endpoint_url)
+{
+    const auto * endpoint_url_pattern_str = R"(http(()|s)://[a-z0-9-]+\.blob\.core\.windows\.net/[a-z0-9-]+)";
```
I think that at least for test purposes should be possible to connect to any url, for example to local docker image like https://hub.docker.com/_/microsoft-azure-blob-storage

ianton-ru reviewed (commented):
I think this feature should be covered with integration tests.

ianton-ru commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,164 @@
+
+off_t ReadBufferFromBlobStorage::getPosition()
+{
+    // TODO: why is it `return offset - available()` in S3?
```
Oh, Kseniia has fastest hand in the wild west.

jkuklis committed:
Addressing first PR remarks

jkuklis commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,164 @@
+        {
+            LOG_INFO(log, "Exception caught during Azure Read for file {} : {}", path, e.Message);
+
+            std::this_thread::sleep_for(sleep_time_with_backoff_milliseconds);
```
Changed.

jkuklis committed:
Starting an integration test for Blob Storage

jkuklis committed:
Working skeleton for Blob Storage integration test

kssenii commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,176 @@
+    if (data_stream == nullptr)
+        throw Exception(ErrorCodes::RECEIVED_EMPTY_DATA, "Null data stream obtained while downloading file {} from Blob Storage", path);
+
+    total_size = data_stream->Length() + offset;
```
Why need to add offset to `data_stream->Length()`?
Also may be makes sense to add tests with both setting values  `remote_filesystem_read_method=read` and  `remote_filesystem_read_method=threadpool`.
Mostly `threadpool` value is more preferable (since it makes reading much much faster), so my be one test for `read` is enough.

jkuklis replied:
I use this `total_size` like `read_until_position` in `nextImpl` (https://github.com/ClickHouse/ClickHouse/pull/31505/files#diff-03e106ad0b2369d201566cd1100a518e1445dfa7b76d3954093525ae93e82f7dR74). As `read_until_position` is not always set, I left it as an universal approach.

kssenii commented on src/IO/WriteBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,73 @@
+namespace DB
+{
+
+// TODO: abstract this function from DiskS3.cpp, from where it was copy-pasted
```
If you  want to remove copy-paste then it is possible to move it to some `RemoteDisksCommon` file.

jkuklis commented on src/Disks/BlobStorage/DiskBlobStorage.cpp:
```diff
@@ -0,0 +1,194 @@
+
+    auto buffer = std::make_unique<WriteBufferFromBlobStorage>(
+        blob_container_client,
+        metadata.remote_fs_root_path + blob_path,
```
Do you think there is a reason to set `remote_fs_root_path` as something else than an empty string in `IDiskRemote` constructor?

kssenii replied:
But it is not set in any other place apart from constructor, but is always used.
It equals `endpoint_url` in your case. But as I see not needed for your case because you create `Azure::Storage::Blobs::BlobContainerClient`, so `endpoint_url` is not needed anymore. So ok to leave empty string. But if it is always empty then probably no need to use `metadata.remote_fs_root_path` anywhere in this disk.

jkuklis commented:
Hello!

Thanks for the review comments, I have addressed some of them and will address the remaining ones soon.

Meanwhile, I started working on the integration tests and got a skeleton for the test to work. It starts a clickhouse-server, registers the Blob Storage Disk and creates a table backed with Blob Storage. All these operations can be done without actually connecting to a Blob Storage endpoint.

To simulate Blob Storage, we can use the Azurite emulator, designed for local Azure Storage development: https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite. It can be run either as a docker container, or using the exec `azurite`, which, if I understand correctly, simply runs the container underneath.

I have experimented with Azurite and I was able to first install it and then connect to it inside the integration tests container using the Python Blob Storage SDK. I have checked both the `azurite` command (installed with `npm`) and the bare docker container, all through `os.system()` invocations in the test implementation. I haven't been able to connect to Azurite through the clickhouse-server yet, but I believe it's a matter of putting these puzzles together in the right way. For now I feel I need to understand a bit more about `ClickHouseCluster` instances to figure it out.

Would you be able to help me with that? Do you have ideas why I am not able to connect to Azurite through clickhouse-server, even though the same setup (endpoints, credentials) works locally on my laptop?

Also, which approach do you prefer, the bare docker or the `azurite` command? What would be a ClickHouse way to run these and add them to dependencies?

I need to modify the registration code to cater for this test environment, especially the authentication and the initial connection parts. Blob Storage has a hierarchy of storage accounts with containers holding blobs, the actual files, 
so far I was connecting directly to a container, but for Azurite I think I will need to connect higher up, to the storage account, as the Azurite storage account comes without any containers.

There are also some remaining doubts and questions in the Further steps section of the PR text, I would be very grateful for some insights into these!

Thanks!

ianton-ru commented:
> the server seems not to see the endpoint (`http://0.0.0.0:10000/devstoreaccount1/first`)

Endpoint IP address 0.0.0.0 looks strange for me. It's a entry network address.
I think that here should be something like 'azurite1' (like 'minio1' for S3 - https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/helpers/cluster.py#L278).

jkuklis commented:
Hey Anton, I needed to make some edits to my previous comment, as I misreported some of my findings, sorry for that.

I have used both `0.0.0.0` and `127.0.0.1` IPs and either just `devstoreaccount1` or `devstoreaccount1/first` (with the container name) in the urls when running my server locally and I could see and connect to all these endpoints.

https://hub.docker.com/_/microsoft-azure-blob-storage - this container is mainly used for IoT Edge devices, and Azurite is a general purpose Blob Storage emulator: https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=docker-hub. It comes with a ready docker image too. Plus, it seems to me that this IoT container is more difficult to configure: https://docs.microsoft.com/en-us/azure/iot-edge/how-to-deploy-blob?view=iotedge-2020-11

ianton-ru commented:
I found another one - https://hub.docker.com/_/microsoft-azure-storage-azurite
I try to run it manually on my laptop, looks like it works. At least I got an HTTP error 400 when try to make request (with wget) from ClickHouse container.

docker-compose.yml:
```
version: "3.7"

services:

  clickhouse1:
    image: yandex/clickhouse-server:latest
    ports:
      - 9000:9000
      - 8123:8123

  azurite1:
    image: mcr.microsoft.com/azure-storage/azurite
    ports:
      - 10000:10000
      - 10001:10001
      - 10002:10002
```
I run it and try
```
$ docker exec -ti azurite_clickhouse1_1 wget http://azurite1:10000/
--2021-11-25 17:03:00--  http://azurite1:10000/
Resolving azurite1 (azurite1)... 192.168.160.2
Connecting to azurite1 (azurite1)|192.168.160.2|:10000... connected.
HTTP request sent, awaiting response... 400 Value for one of the query parameters specified in the request URI is invalid.
2021-11-25 17:03:00 ERROR 400: Value for one of the query parameters specified in the request URI is invalid..
```

jkuklis committed:
Running Azurite docker in the integration test for Blob Storage support

jkuklis commented:
I added the `with_azurite` setting to `ClickHouseCluster` method `add_instance`, which runs additional docker container for Azurite (image `mcr.microsoft.com/azure-storage/azurite`). I run my tests locally with `pytest`* (not in the docker with `./runner`, as I have this `NETWORK_ERROR` there) and I can connect to the created container in `wait_azurite_to_start()` in cluster.py. However, clickhouse-server still seems not to see the endpoints exposed by the container, I don't see any sign of them connecting, the sockets time out.

I tried to run Azurite directly from the container holding the clickhouse-server instance, but I wasn't able to connect to the net to do `apt-update` there to try to install packages like `docker`. The only preinstalled tool I have found there was `wget`, but again, I wasn't able to connect to the net there.

Is some proxy needed for Azurite? How to make the endpoints exposed by the docker visible to the clickhouse-server instance?

* At least it used to be the case, I have just got this error with IP `172.22.0.3:9000` with `pytest`, I need to see what could have changed.

kssenii commented on tests/integration/test_merge_tree_blob_storage/configs/config.d/storage_conf.xml:
```diff
@@ -0,0 +1,20 @@
+        <disks>
+            <blob_storage_disk>
+                <type>blob_storage</type>
+                <endpoint>?</endpoint>
```
Why `?` is used for endpoint?
Did you try to put something like here `http://azurite1:10000/`?

>  I run my tests locally with pytest* (not in the docker with ./runner, as I have this NETWORK_ERROR there) 

btw in our CI tests are run with runner, not with pytest. So need to make it work locally with runner.

jkuklis replied:
I modified the Disk registration code on the side and didn't use this endpoint for this development phase, eventually it should be something like `http://azurite1:10000/`. I am still figuring out how to connect different pieces and make them work, I have these mentioned issues with connecting different containers. I am also still working out the `NETWORK_ERROR` problem, but I don't have a clear suspect.

kssenii replied:
> I modified the Disk registration code on the side and didn't use this endpoint for this development phase, eventually it should be something like http://azurite1:10000/

Ok, but it could be more convenient if you added some endpoint (at least this one ` http://azurite1:10000/`, it looks correct) to config so it is used when disk is registered and some test in CI started. So it would be easier to understand by CI logs what errors do you get and to checkout to you branch and also try to check how to make it work.

kssenii replied:
Can you please fix build (cmake issues)? Because it does not build now as is. It is easier for me to checkout to your branch and check myself what is going on with your tests. I have added really a lot of integration tests myself for external engines/storages including those which had no support before, so your case should have little difference.

jkuklis committed:
Adding random postfix to blob names, needs code cleanup

jkuklis committed:
Removing file checks after changing blob filenames

jkuklis committed:
Rebased on master

jkuklis committed:
Removing spurious '/' character

jkuklis committed:
Fixing submodule update for clickhouse compilation in the fast test

jkuklis committed:
Add missing dependencies for azure for fast test

kssenii commented on docker/test/fasttest/run.sh:
```diff
@@ -187,6 +187,8 @@ function clone_submodules
             contrib/NuRaft
             contrib/azure
+            contrib/boringssl
+            contrib/curl
```
oh, well it is better not to enable building extra contribs for fast test. Better to do it the same as for other libraries -- check in `cmake/find/block_storage/cmake` contains for example file `${ClickHouse_SOURCE_DIR}/contrib/azure/cmake-modules/AzureTransportAdapters.cmake` and if not -- `set(USE_AZURE_BLOB_STORAGE 0)`.
Sorry did not take a look at the reason fast test failed before now.

jkuklis committed:
Reverting submodule changes to fast test docker and disabling Blob Storage support when files missing

jkuklis committed:
Code cleanup addressing PR remarks

jkuklis committed:
Enable connecting directly to storage accounts and creating new containers

jkuklis committed:
Adding endpoint validation

jkuklis committed:
Enable authentication via StorageSharedKeyCredential

kssenii commented on tests/integration/test_merge_tree_blob_storage/configs/config.d/storage_conf.xml:
```diff
@@ -0,0 +1,24 @@
+                <skip_access_check>true</skip_access_check>
+                <!-- default credentials for Azurite storage account -->
+                <account_name>devstoreaccount1</account_name>
+                <account_key>Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==</account_key>
```
With this configuration, when I run  integration tests, I get 
```
2021.12.01 22:00:59.151040 [ 59 ] {} <Error> void DB::SystemLog<DB::TraceLogElement>::flushImpl(const std::vector<LogElement> &, uint64_t) [LogElement = DB::TraceLogElement]: std::exception. Code: 1001, type: Azure::Storage::StorageException, e.what() = 400 Invalid storage account.
Invalid storage account.
RequestId:f36ca285-7d36-4c5a-a8a5-c9075399882f
Time:2021-12-01T19:00:57.849Z
0. /home/kssenii/workspace/CH-3/build-debug/../contrib/libcxx/include/exception:133: std::exception::capture() @ 0x15270e22 in /usr/bin/clickhouse
1. /home/kssenii/workspace/CH-3/build-debug/../contrib/libcxx/include/exception:111: std::exception::exception() @ 0x15270df3 in /usr/bin/clickhouse
2. /home/kssenii/workspace/CH-3/build-debug/../contrib/libcxx/src/support/runtime/stdexcept_default.ipp:33: std::runtime_error::runtime_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x2a44ec9d in /usr/bin/clickhouse
3. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/core/azure-core/inc/azure/core/exception.hpp:78: Azure::Core::RequestFailedException::RequestFailedException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x26fe74c1 in /usr/bin/clickhouse
4. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/storage/azure-storage-common/inc/azure/storage/common/storage_exception.hpp:26: Azure::Storage::StorageException::StorageException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x270b9461 in /usr/bin/clickhouse
5. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/storage/azure-storage-common/src/storage_exception.cpp:144: Azure::Storage::StorageException::CreateFromResponse(std::__1::unique_ptr<Azure::Core::Http::RawResponse, std::__1::default_delete<Azure::Core::Http::RawResponse> >) @ 0x270b7609 in /usr/bin/clickhouse
6. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:4760: Azure::Storage::Blobs::_detail::BlobRestClient::BlobContainer::Create(Azure::Core::Http::_internal::HttpPipeline&, Azure::Core::Url const&, Azure::Storage::Blobs::_detail::BlobRestClient::BlobContainer::CreateBlobContainerOptions const&, Azure::Core::Context const&) @ 0x27042ce7 in /usr/bin/clickhouse
7. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/storage/azure-storage-blobs/src/blob_container_client.cpp:147: Azure::Storage::Blobs::BlobContainerClient::Create(Azure::Storage::Blobs::CreateBlobContainerOptions const&, Azure::Core::Context const&) const @ 0x270409c8 in /usr/bin/clickhouse
8. /home/kssenii/workspace/CH-3/build-debug/../contrib/azure/sdk/storage/azure-storage-blobs/src/blob_service_client.cpp:235: Azure::Storage::Blobs::BlobServiceClient::CreateBlobContainer(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Azure::Storage::Blobs::CreateBlobContainerOptions const&, Azure::Core::Context const&) const @ 0x27064886 in /usr/bin/clickhouse
9. /home/kssenii/workspace/CH-3/build-debug/../src/Disks/BlobStorage/registerDiskBlobStorage.cpp:177: DB::getBlobContainerClient(DB::BlobStorageEndpoint const&, Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x215186fe in /usr/bin/clickhouse

```
I don't know azure blob storage and how to make a valid account.
But looks like it means it gets connected ok.

jkuklis replied:
I am about to check these connection issues with `Invalid storage account`, I am not sure yet that this is indeed a sign of connecting with Azurite. I am surprised it doesn't fail at `ListBlobContainers`, src/Disks/BlobStorage/registerDiskBlobStorage.cpp:171. I am also about to check the logs in the `_instances` directory.

I didn't notice this missing exception before, I think it is the reason why I thought I managed to connect to Azurite in the test.

I updated the Dockerfile to install `azure-storage-blob`.

I think I don't have the access rights to put the library in `ClickHouse-extras`, is that correct?

I have a new problem with running the tests with `./runner`, I get `FileNotFoundError: [Errno 2] No such file or directory: '/compose/docker_compose_azurite.yml'` that I will investigate now as well.

kssenii replied:
> rights to put the library in ClickHouse-extras, is that correct?

Yes, I can fork azure blob storage libtary there and you can make a pull request there if you want. Not necessary though. 

> I have a new problem with running the tests with ./runner

btw, just in case you missed this don't forget to do this https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/README.md?plain=1#L73

> I have a new problem with running the tests with ./runner, I get FileNotFoundError: [Errno 2] No such file or directory: '/compose/docker_compose_azurite.yml' that I will investigate now as well.

Yes, you need to go to `docker/test/integration/runner/` and execute command you will see in the comment in the first line of Dockerfile there.

jkuklis replied:
I edited the comment above in the meantime, posting the update here:

Rebuilding the image locally fails - when I do it, there are issues with apt-update, it looks like there is no connection to the network within the container, I get e.g. `Temporary failure resolving 'ru.archive.ubuntu.com`. I wasn't able to modify the image and run it with ./runner, there were some errors. I needed to create the file from scratch using `os.system` in the test.py code.

jkuklis replied:
But I got the `Invalid storage account` error as well, I can investigate further now.

kssenii replied:
> Rebuilding the image locally fails - when I do it, there are issues with apt-update, it looks like there is no connection to the network within the container, I get e.g. Temporary failure resolving 'ru.archive.ubuntu.com

oh, I recently started to get the same error. It gets fixed by `sudo service docker restart`

jkuklis committed:
Installing azure-storage-blob in the integration tests docker, raising exception when not connected to Azurite, changing ports in the Azurite docker

jkuklis committed:
Solved Invalid storage account error with connection strings

jkuklis committed:
Comming back to StorageSharedKeyCredential for the integration test, get 400 Bad requests from Azurite when trying to create a container now

jkuklis committed:
Connecting to Azurite and invoking simple INSERTs works in the integration test

jkuklis committed:
First integration tests, the one for merging fails for the moment

jkuklis committed:
Merging test works correctly

jkuklis committed:
More integration tests, all pass, at least usually

jkuklis committed:
Minor tweaks to the test, disabling another slow one

jkuklis committed:
Abstracting metadata creation for remote disks registrations, putting Blob Storage auth in a separate file

jkuklis committed:
Committing missing files

jkuklis committed:
Abstracting utility functions from main integration tests for remote disks

jkuklis committed:
Simple check for files in cloud

jkuklis commented on src/Disks/BlobStorage/BlobStorageAuth.cpp:
```diff
@@ -0,0 +1,138 @@
+
+    auto blob_service_client = getBlobStorageClientWithAuth<BlobServiceClient>(endpoint.storage_account_url, container_name, config, config_prefix);
+
+    if (!endpoint.container_already_exists.has_value())
```
Here I decided to only list containers if the user doesn't specify `container_already_exists` at all. If the user specifies `container_already_exists` to `true`, then we get to line 115, and I think an error should be thrown if the container doesn't exist, I will double check. In any case, we won't create a new container if user specifies `container_already_exists` to `true`.

kssenii replied:
ah, yes, misread this case `if (!endpoint.container_already_exists.has_value())` like `if (endpoint.container_already_exists.has_value())`

jkuklis commented on src/IO/ReadBufferFromBlobStorage.cpp:
```diff
@@ -0,0 +1,174 @@
+    size_t tmp_buffer_size_,
+    bool use_external_buffer_,
+    size_t read_until_position_)
+    // TODO: shall this notation be used in all constructors?
```
I meant the notation with `:` and `,` at the beginning of lines, rather than at the end.

jkuklis committed:
Cache test, removing spurious imports

jkuklis committed:
Zero replication support test

jkuklis committed:
Update azure submodule

jkuklis committed:
Trying to solve compilation issue for a GCC build

jkuklis commented:
Hello!

The PR has moved forward, here's a summary of the status:
* DiskBlobStorage is functional, it implements all the main capabilites of a remote disk,
* in local tests everything seems to work as expected, I am not able to trigger any errors by fiddling with the client,
* integration tests are implemented and they pass (at least most of the time),
* another test with production data has run for 10 days now, no shard has experienced a fatal crash, there have only been sporadic `FILE_DOESNT_EXIST` errors when looking for Blob Storage files,
* the code is already quite stable, my remaining questions mostly resolve around code sharing for remote disks.

I think DiskBlobStorage is already well-rounded now that I have added other means of authentication.

For the integration tests, it happened to me that an error for an already existing container popped up with `container_already_exists` set to `false`, as if the Azurite was already populated, maybe when I run two integration tests immediately one after another. I am not sure if that is a problem, it doesn't happen on a regular basis. If necessary, I can remove the `container_already_exists` line from the configuration.

Other than that, I disabled two tests by default, as they were sometimes too slow for the pipeline. Especially the `test_restart_during_load` can take several minutes to run when run together with other tests, on its own it generally takes less than a minute. Maybe I need to clean the server after each test.

I added an utility.py file sharing some common functionalities for the Blob Storage, S3, HDFS tests, and also other ones. I don't wanna make this a refactoring PR, I only made some modifications to files test_merge_tree_s3/hdfs.py to show what I have in mind. My impression is that there are a couple more such utilities spread across different integration tests that could be unified and shared.

To a lesser extent this issue also affects the current implementation of remote disks, I added the RemoteDisksCommon.h/cpp files to mitigate it.

The production test seems to confirm that Blob Storage support generally works well, though probably there are some minor parts to improve. When I check with the `FINAL` modifier, I see that the data in the test and the reference shards matches, actually seemingly no data has been missed during these 10 days. A couple of times these `FILE_DOESNT_EXIST` errors started popping up (usually for all the files in the given insert/merge directory at once), but they are not fatal. It looks as if the number of rows in general is a bit larger in the test shards, I think some row updates haven't been processed correctly (hence the `FILE_DOESNT_EXIST` errors) and there are some spurious rows with +1, -1 signs. Still, this issue isn't as noticeable as it was before, the extra rows constitute maybe ~1% of the whole data.

Any suggestions for the further steps?

Thanks, especially thanks a lot @kssenii, for monitoring this PR!

jkuklis committed:
Fixing variable shadow error

jkuklis committed:
Style corrections

kssenii reviewed (approved):
LGTM. We can merge once CI tests finish successfuly.

kssenii commented:
About your unanswered questions in PR description:

> What is the main reason for having a POCO client for S3, considering AWS already provides HTTP clients using e.g. libcurl?
What is the purpose of the proxy configuration part for S3?

I think better to ask @ianton-ru.

Other questions as I see resolved or covered in code well.

kssenii commented:
Integration tests fail because `[Errno 2] No such file or directory: '/compose/docker_compose_azurite.yml'`, i.e. docker image did not get updated, fixed by https://github.com/ClickHouse/ClickHouse/pull/32438. You can merge changes from that PR into yours to let integration tests start in your PR earlier that that PR is merged.

`AST fuzzer (debug, actions) — Task failed: $?=132` also fails, but the details link is strange for some reason (https://s3.amazonaws.com/clickhouse-test-reports/31505/d023ad16ed171c8f6192cc2c859a1a5a7c4adc9f/fuzzer_astfuzzerdebug,actions//server.log). Good that I already stepped into this `gmtime` when building your branch in debug mode, so this report tells about forbidden `gmtime`  function. Looks like will need to fix this after all (though actually very strange that it complains, because we have forbidden function in hdfs library too, but it does not complain until you run integration tests with debug build, which are not currently run in debug in CI). I forked azure library to clickhouse-extras (https://github.com/ClickHouse-Extras/azure-sdk-for-cpp), you can make a PR with substitution..

jkuklis commented:
I added a PR for ClickHouse-Extras Azure SDK, but I think that another function, `localeconv`, might also cause problems.

kssenii commented:
> I added a PR for ClickHouse-Extras Azure SDK, but I think that another function, localeconv, might also cause problems.

Yes: https://github.com/ClickHouse/ClickHouse/blob/f7711c8e426476e662e7a78430d7cbaf91dc7526/base/harmful/harmful.c#L113

may be `gmtime_r`? It is not in the list..

jkuklis committed:
Updating azure submodule with ClickHouse-Extras fork

jkuklis commented:
I forgot to remove TODOs in the last commit, will do it once the tests are finished.

jkuklis commented:
The **Integration tests flaky check (asan, actions)** test still fails on all the Blob Storage tests, with warnings liks `Cannot find image clickhouse/integration-tests-runner in params list {'yandex/clickhouse-integration-tests-runner': 'latest' ...`, despite updating the CI script. This #32438 PR is already merged, I will rebase on master when more tests will have finished.

jkuklis committed:
Removing TODO and NOTE marks used mainly as hooks for review in the PR

kssenii commented:
This should finally fix https://github.com/ClickHouse/ClickHouse/pull/32494.

jkuklis commented:
I think the remaining tests require approval @kssenii, thanks in advance!

kssenii commented:
> Hm, some further problems with CI/CD it seems.

Yes... But we can ignore `functional tests` checks, for this PR only matter integration tests, build checks, pvs and style check.
For now I see `Integration tests flaky check (asan, actions)` and in `integration_run_flaky_0.log` report there is 
```
        if (self.process.returncode != 0 or stderr) and not self.ignore_error:
>           raise QueryRuntimeException(
                'Client failed! Return code: {}, stderr: {}'.format(self.process.returncode, stderr), self.process.returncode, stderr)
E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 21.13.1):
E           Code: 1001. DB::Exception: Received from 172.16.4.3:9000. DB::Exception: Azure::Core::Http::TransportException: Connection was closed by the server while trying to read a response. 
```

jkuklis commented:
This test always passes quickly on my laptop, I don't see what could be wrong with it, shall I just push an empty commit to trigger the tests again?

kssenii commented:
> This test always passes quickly on my laptop, I don't see what could be wrong with it, shall I just push an empty commit to trigger the tests again?

Yes, may be it will help. I looked at  `integration_run_flaky_0.tar.gz` in `docker.log` and `clickhouse-server.log` and compared time of logging and looks like docker.log stopped earlier, though AFAIK if it was CI issue then the error in check would more distinctly point to that (not DB::Exception).

jkuklis committed:
Changing order of tests to check if that affects failing tests in the CI/CD

jkuklis commented:
I changed the order of the tests to see if it is only the last Blob Storage test in the pipeline, which fails, but now the different **Integration tests (asan, actions)** reports show `Invalid check_status.tsv`, and this file doesn't show up in the _Details_ link.

kssenii commented:
> reports show Invalid check_status.tsv

It broken and then was fixed in master some little time ago. Next tests run should be ok.

jkuklis committed:
Removing any extra work and correcting the cache path in the cache test

jkuklis commented:
I don't know how to debug this cache test issue.

The cache test case fails in the CI/CD pipeline every time it is run, and not only as the last Blob Storage test in the group, but I it passes both on my laptop and my Azure dev VM. There was a mistake in the path there, it was `/var/.../blob_storage/cache/` rather than `/var/.../blob_storage_disk/cache/` (with `blob_storage_disk` like in the storage configuration), but I don't think it was causing the failure. I pushed a commit fixing this path and removing logs manipulations from the test case, we'll see if it helps with the CI/CD.

As I can't reproduce the issue locally, and the CI/CD pipeline takes several hours to finish this test, it is hard for me to debug it. Could you help me fix it?

kssenii commented:
Let's see one last time how it goes, if still fail, I'm ok to add `@pytest.mark.skip` for this test.
I see this test always passes well in simple integration tests check and fails in flacky tests check only from the first run. Usually when this happens the problem is because tests order is randomized and some trash remained after previous tests.

jkuklis commented:
Hm, from what I see the order of the test cases is always the same if these test cases are in the same test file, I could also put the cache test as the first test in the file. One test which could maybe affect the behaviour of the following tests is `test_apply_new_settings`, but it seems to pass without problems.

kssenii commented:
> Hm, from what I see the order of the test cases is always the same if these test cases are in the same test file

In our CI the order within the same test file is randomized always, this is for sure.

jkuklis committed:
Cache test as the first test in the file

jkuklis commented:
In the https://s3.amazonaws.com/clickhouse-test-reports/31505/38e35a0ef537d6364f0726adb6a7b72dc049e53e/integration_tests_flaky_check__asan__actions_/integration_run_flaky_0.log and https://s3.amazonaws.com/clickhouse-test-reports/31505/60e06ca34c2186f2f34e0a0bc8f39994cda72ca8/integration_tests_flaky_check__asan__actions_/integration_run_flaky_0.log the tests are listed in the same order, I think this should always be the case with `pytest`, at least I don't see any sign that there is some randomization in the order of these integration tests.

jkuklis commented:
The flaky integration test passed for the last commit, putting the cache wipe test as the first test in the filef seemingly helped. Is this setup fine for you, @kssenii ?
Other than that, I think the PR is ready to merge, I can maybe rebase the last time on master if necessary.

kssenii commented:
Integration tests ok:
`Integration tests flaky check (asan, actions) — No flaky tests: error: 0, passed: 33`
test failures not related:
`Integration tests (asan, actions) [2/3] — fail: 1` -- test_storage_kerberized_kafka/test.py::test_kafka_json_as_string_no_kdc
`Stateless tests (release, wide parts enabled, actions) ` -- 01814_distributed_push_down_limit
`Stress test (thread, actions) — Test script failed`

jkuklis commented:
Thanks a lot for help with reviews @kssenii and @ianton-ru !

I hope this PR will prove useful ^^
